MapleServer2/PacketHandlers/Game/InsigniaHandler.cs

```diff
@@ -14,15 +17,80 @@ public class InsigniaHandler : GamePacketHandler
 
         public override void Handle(GameSession session, PacketReader packet)
         {
+
+            ProcessInsignias(session);
+
             short insigniaId = packet.ReadShort();
 
             if (insigniaId < 0)
             {
                 return;
             }
 
-            session.Player.InsigniaId = insigniaId;
-            session.FieldManager.BroadcastPacket(InsigniaPacket.UpdateInsignia(session, insigniaId, true));
+            if (session.Player.AvaliableInsignias.Contains(insigniaId))
+            {
+                session.Player.InsigniaId = insigniaId;
+            }
+            else
+            {
+                insigniaId = 0;
+            }
+
+            session.FieldManager.BroadcastPacket(InsigniaPacket.UpdateInsignia(session, insigniaId, insigniaId != 0));
+        }
+
+        private void ProcessInsignias(GameSession session)
+        {
+            Player player = session.Player;
+            int[] trophy = player.Trophy;
+            List<int> avaliableTitles = player.AvaliableTitles;
+            List<short> tempInsignias = new List<short> { };
+            Dictionary<ItemSlot, Item> equips = player.Equips;
+
+            Dictionary<int, short> titlesThatGiveInsignias = new Dictionary<int, short> {
```

Comment:
I'd move this into a constant outside of here.

Reply:
Sure!